File: ai-lien/shutdown.py
# ---------------------------------------------------------------------------
# Graceful shutdown — x-tca-lifecycle
#
# Contract: contracts/openapi/ai-lien.yaml x-tca-lifecycle
# Sequence: stop-accepting-connections → drain-in-flight-requests → exit-0
#
# terminationGracePeriodSeconds in Helm must exceed drain-timeout-seconds.
# ---------------------------------------------------------------------------

import logging
import signal
import sys
from http.server import HTTPServer

log = logging.getLogger('ai-lien')

DRAIN_TIMEOUT_SECONDS = 15


def register_shutdown(server: HTTPServer) -> None:
    """Wire SIGTERM and SIGINT handlers onto the running server.
    Call this immediately after server construction, before serve_forever().
    """

    def _handler(signum: int, _frame) -> None:
        sig_name = signal.Signals(signum).name
        log.info(f'Received {sig_name} — starting graceful shutdown')

        # Step 1+2: Stop accepting connections and drain in-flight requests.
        # shutdown() signals serve_forever() to exit after the current request.
        # server_close() releases the socket.
        try:
            server.shutdown()
            server.server_close()
        except Exception as e:
            log.error(f'Server shutdown error: {e}')

        log.info('Shutdown complete')
        sys.exit(0)

    signal.signal(signal.SIGTERM, _handler)
    signal.signal(signal.SIGINT,  _handler)

File: signal-clearance/src/shutdown.ts
// ---------------------------------------------------------------------------
// Graceful shutdown — x-tca-lifecycle
//
// Contract: contracts/openapi/signal-clearance.yaml x-tca-lifecycle
// Sequence: stop-accepting-connections → drain-in-flight-requests →
//           disconnect-redis → exit-0
//
// terminationGracePeriodSeconds in Helm must exceed drain-timeout-seconds.
// ---------------------------------------------------------------------------

import https from 'https';
import { disconnectRedis } from './sessions.js';

const DRAIN_TIMEOUT_MS = 15_000;

// registerShutdown wires SIGTERM and SIGINT handlers onto the running server.
// Call this immediately after server.listen() in main.
export function registerShutdown(server: https.Server): void {
  const handler = async (signal: string) => {
    console.log(`[signal-clearance] Received ${signal} — starting graceful shutdown`);

    // Step 1+2: Stop accepting new connections and drain in-flight requests.
    // Node.js server.close() stops the listener; existing connections finish.
    const closePromise = new Promise<void>((resolve) => {
      server.close(() => resolve());
    });

    // Enforce drain timeout — don't hang forever if a connection is stuck.
    const timeout = setTimeout(() => {
      console.log('[signal-clearance] Drain timeout reached — forcing shutdown');
      process.exit(1);
    }, DRAIN_TIMEOUT_MS);

    try {
      await closePromise;
      clearTimeout(timeout);
    } catch (err) {
      console.error('[signal-clearance] Server close error:', err);
    }

    // Step 3: Disconnect Redis — all session reads have completed.
    try {
      await disconnectRedis();
    } catch (err) {
      console.error('[signal-clearance] Redis disconnect error:', err);
    }

    console.log('[signal-clearance] Shutdown complete');
    process.exit(0);
  };

  process.on('SIGTERM', () => handler('SIGTERM'));
  process.on('SIGINT',  () => handler('SIGINT'));
}
